client/zecale/cli/zecale_check_batch.py

```diff
@@ -0,0 +1,35 @@
+# Copyright (c) 2015-2020 Clearmatics Technologies Ltd
+#
+# SPDX-License-Identifier: LGPL-3.0+
+
+import json
+from click import argument, option, command, pass_context, Context, ClickException
+
+
+@command()
+@argument("batch-file")
+@option(
+    "--batch-size",
+    type=int,
+    required=True,
+    help="Number of nested proofs per batch")
+@pass_context
+def check_batch(ctx: Context, batch_file: str, batch_size: int) -> None:
+    """
+    Exit with error if result is not 1 for any nested proof
+    """
+    with open(batch_file, "r") as batch_proof_f:
+        batch_proof = json.load(batch_proof_f)
+
+    # Attempt to automatically detect the vk_hash param at position 0
+    inputs = batch_proof["inputs"]
+    batch_offset = len(inputs) % batch_size
+    inputs_per_batch = int((len(inputs) - batch_offset) / batch_size)
+    print(f"inputs=\n{inputs}")
+
+    for i in range(batch_size):
+        result_idx = batch_offset + (inputs_per_batch * (i + 1)) - 1
+        result = int(inputs[result_idx][2:], 16)
+        print(f"result[{i}]={result}")
+        if result != 1:
+            raise ClickException("nested proof judged as invalid")
```

Comment:
I don't understand the purpose of this subcommand. What is it made for?

What does the `batch_file` contain? The set of nested proofs ("the batch") or the "wrapping proof" (which is loaded in memory in the `batch_proof` variable?)?

Reply:
The `get_batch` command returns an aggregated proof, inputs and parameters for a batch of nested transactions.
This command is used to check the results of the nested proofs.  For now it only checks that all nested proofs verify correctly, but eventually it could later be extended to support asserting failure as well.

(This allows clients to ensure that their nested transactions were judged as valid as expected - if anything in the pipeline goes wrong and the nested proof is judged as invalid, the client has no other way to tell because a valid batch is always returned.)

Reply:
Is that used for Debug purposes? Where is the data written in the `batch_file`?
Why not use `GenerateAggregateProof` (see: https://github.com/clearmatics/zecale/blob/master/api/aggregator.proto#L24-L32) that returns the aggregation proof and the related inputs? (Note however that as written in the comments on the code excerpt linked, this endpoint is only useful for some debugging/manual triggering for now. In the end, we may want the aggregator to be ran on another machine which will send the aggregation proof on chain as soon as it is generated, which itself can be automatically triggered when a threshold of tx_to_aggregate have been pushed on the application pool)

Reply:
The batch is received in `zecale get_batch` (which naturally uses the GenerateAggregateProof rcp call) .  `check-batch` can then be used to check that batch.  I would not describe it as debug functionality, as such.  More a way for the receiver of the batch to interrogate it.

Eventually, this check will probably happen via the blockchain and the zecale contract will emit events and clients will need a `sync` operation instead of `get-batch` and `check-batch`.  But for now, until we have this infrastructure, the client uses `get-batch` and `check-batch`, along with some assumptions about the aggregator, in order to run the full pipeline.

Reply:
> The batch is received in zecale get_batch (which naturally uses the GenerateAggregateProof rcp call) . check-batch can then be used to check that batch.

I see. Thanks.
Out of curiosity, if this makes the assumption that `zecale get_batch` was ran before, why has it been moved in its own command and not as a flag - `zecale get_batch --check`?

> I would not describe it as debug functionality, as such. More a way for the receiver of the batch to interrogate it [...]
> But for now, until we have this infrastructure, the client uses get-batch and check-batch, along with some assumptions about the aggregator, in order to run the full pipeline.

I see. So what happens if `zecale check-batch` fails? Do you abort the whole pipeline (as in https://github.com/clearmatics/zecale/blob/bw6-groth16-verifier/scripts/test_client#L39-L44)?
If so, that's not quite desired. We still want to process (i.e. we don't want to "censor") tx with invalid nested proofs, right? The same way Zeth tx containing an invalid proof are processed on chain. In the context of Zecale, their instances just won't be "dispatched" to the base app contract.

Reply:
> Out of curiosity, if this makes the assumption that zecale get_batch was ran before, why has it been moved in its own command and not as a flag - zecale get_batch --check

It only makes the assumption that given file includes a batch.  This is just for versatility.  `get-batch` retrieves a batch. `check-batch` performs some check on it (currently very trivial, but we may want other checks - currently there is no way to even identify which tx's are in the batch), `submit-batch` submits it to the dispatcher contract.

> I see. So you don't plan to use this check-batch command in the end state, right?

I would expect many of the current commands to change or disappear over time.  The specifics will depend purely on the configurations we want to support.  In a configuration where batches are only ever produced by 3rd party entities then the client will not need `get-batch`, `check-batch` or `submit-batch`.

> Aborting the whole pipeline if a nested proof is invalid (as "emulated" in https://github.com/clearmatics/zecale/blob/bw6-groth16-verifier/scripts/test_client#L39-L44) wouldn't be desired as we still want to process (i.e. we don't want to "censor") tx with invalid nested proofs. The same way Zeth tx containing an invalid proof are processed on chain. In the context of Zecale, their instances just won't be "dispatched" to the base app contract.

I think there is a misunderstanding here.  Aborting the pipeline is definitely not the intention of `check-batch`.  This command performs a check and exits with 0 or 1 based on whether the condition holds - whether the batch is still submitted to the dispatcher is entirely up to the caller.

`test_client` is a test script which creates and submits batched, using `check-batch` to confirm assumptions about the nested proof validity. So the script exits with error if its assumptions do not hold, but in general I would not expect the entire batch to be discarded based on this.  (There was actually a bug where nested proof validation was failing, but without any tools to look at the batch this was not being caught until we checked the application state.)  Depending on the configuration of the system, many checks by the client will probably be done via log events, but as mentioned I think that requires some more changes first.

Reply:
> I think there is a misunderstanding here. Aborting the pipeline is definitely not the intention of check-batch. This command performs a check and exits with 0 or 1 based on whether the condition holds - whether the batch is still submitted to the dispatcher is entirely up to the caller.

I misunderstood the purpose indeed! 
Ok that makes sense to me. Let's iterate all this going forward :)

Reply:
> I think there is a misunderstanding here. Aborting the pipeline is definitely not the intention of check-batch. This command performs a check and exits with 0 or 1 based on whether the condition holds - whether the batch is still submitted to the dispatcher is entirely up to the caller.

I see. That's what I had in mind by "debug purpose" (i.e. it does not alter the overall execution flow but is rather used to carry out some sanity checks on the received data). That makes sense then :)
Let's iterate all this going forward :)